File: ihchat/php/app/Atendimento/Metricas.php
<?php
declare(strict_types=1);

namespace IHchat\Atendimento;

use IHchat\Banco\Banco;
use IHchat\Nucleo\Datas;
use IHchat\Nucleo\Json;

/** Números do atendimento para o topo do painel (app/api/metricas.py). */
final class Metricas
{
    /**
     * MetricasSaida: {abertas, pendentes, resolvidas_hoje, sem_atendente,
     * mensagens_hoje, por_canal: {nome: total}, tempo_medio_primeira_resposta_seg}
     *
     * "Hoje" é o dia UTC, como no Python.
     *
     * $recorte: condição SQL sobre o alias "c" de conversas (com parâmetros)
     * que limita os números às conversas de um setor (Visibilidade); null =
     * o atendimento inteiro.
     *
     * @param array{0: string, 1: list<mixed>}|null $recorte
     * @return array<string, mixed>
     */
    public static function resumo(?array $recorte = null): array
    {
        $inicio = Datas::paraBanco(Datas::agora()->setTime(0, 0));
        [$dentro, $pr] = $recorte ?? ['1 = 1', []];
        $contar = static fn (string $onde, array $p = []): int
            => (int) Banco::valor("SELECT COUNT(c.id) FROM conversas c WHERE {$onde} AND {$dentro}", [...$p, ...$pr]);

        $porCanal = [];
        foreach (Banco::todos(
            "SELECT ca.nome, COUNT(c.id) AS total FROM canais ca JOIN conversas c ON c.canal_id = ca.id
             WHERE c.status <> 'resolvida' AND {$dentro} GROUP BY ca.nome ORDER BY ca.nome",
            $pr
        ) as $linha) {
            $porCanal[(string) $linha['nome']] = (int) $linha['total'];
        }

        // tempo até a primeira resposta, só das conversas já respondidas
        $soma = 0.0;
        $quantas = 0;
        foreach (Banco::todos(
            "SELECT c.criada_em, c.primeira_resposta_em FROM conversas c WHERE c.primeira_resposta_em IS NOT NULL AND {$dentro}",
            $pr
        ) as $linha) {
            $criada = Datas::doBanco((string) $linha['criada_em']);
            $resposta = Datas::doBanco((string) $linha['primeira_resposta_em']);
            if ($criada === null || $resposta === null) {
                continue;
            }
            $soma += ((float) $resposta->format('U.u')) - ((float) $criada->format('U.u'));
            $quantas++;
        }

        return [
            'abertas' => $contar("c.status = 'aberta'"),
            'pendentes' => $contar("c.status = 'pendente'"),
            'resolvidas_hoje' => $contar("c.status = 'resolvida' AND c.resolvida_em >= ?", [$inicio]),
            'sem_atendente' => $contar("c.atendente_id IS NULL AND c.status <> 'resolvida'"),
            'mensagens_hoje' => $recorte === null
                ? (int) Banco::valor('SELECT COUNT(id) FROM mensagens WHERE criada_em >= ?', [$inicio])
                : (int) Banco::valor(
                    "SELECT COUNT(m.id) FROM mensagens m JOIN conversas c ON c.id = m.conversa_id WHERE m.criada_em >= ? AND {$dentro}",
                    [$inicio, ...$pr]
                ),
            'por_canal' => Json::objeto($porCanal),
            'tempo_medio_primeira_resposta_seg' => $quantas === 0 ? null : round($soma / $quantas, 1),
        ];
    }
}

File: ihchat/app/api/metricas.py
"""Numeros do atendimento para o topo do painel."""
from __future__ import annotations

from datetime import datetime, time, timezone

from sqlalchemy import func, select

from fastapi import APIRouter

from .. import permissoes as perm
from ..dependencias import AtendenteAtual, Sessao
from ..models import Canal, Conversa, Mensagem, StatusConversa, agora
from ..schemas import MetricasSaida
from ..util import garantir_utc

rotas = APIRouter(prefix="/api/metricas", tags=["metricas"])


def _inicio_do_dia() -> datetime:
    return datetime.combine(agora().date(), time.min, tzinfo=timezone.utc)


@rotas.get("/resumo", response_model=MetricasSaida)
def resumo(sessao: Sessao, atual: AtendenteAtual) -> MetricasSaida:
    """metricas.ver_todas: o atendimento inteiro; metricas.ver_setor: o
    recorte do próprio setor (as conversas que o setor vê); sem nenhuma, 403."""
    from ..servicos import visibilidade

    recorte = None
    if not perm.tem(atual, "metricas.ver_todas"):
        perm.exigir(atual, "metricas.ver_setor")
        recorte = visibilidade.condicao(atual, como_setor=True)
    dentro = [recorte] if recorte is not None else []
    inicio = _inicio_do_dia()

    def contar(*condicoes) -> int:
        return sessao.scalar(select(func.count(Conversa.id)).where(*condicoes, *dentro)) or 0

    por_canal = {
        nome: total
        for nome, total in sessao.execute(
            select(Canal.nome, func.count(Conversa.id))
            .join(Conversa, Conversa.canal_id == Canal.id)
            .where(Conversa.status != StatusConversa.RESOLVIDA.value, *dentro)
            .group_by(Canal.nome)
        )
    }

    consulta_mensagens = select(func.count(Mensagem.id)).where(Mensagem.criada_em >= inicio)
    if dentro:
        consulta_mensagens = consulta_mensagens.join(Conversa, Conversa.id == Mensagem.conversa_id).where(*dentro)
    mensagens_hoje = sessao.scalar(consulta_mensagens) or 0

    # tempo ate a primeira resposta, so das conversas ja respondidas
    tempos = sessao.execute(
        select(Conversa.criada_em, Conversa.primeira_resposta_em).where(
            Conversa.primeira_resposta_em.is_not(None), *dentro
        )
    ).all()
    media = None
    if tempos:
        segundos = [
            (garantir_utc(resposta) - garantir_utc(criada)).total_seconds()
            for criada, resposta in tempos
        ]
        media = round(sum(segundos) / len(segundos), 1)

    return MetricasSaida(
        abertas=contar(Conversa.status == StatusConversa.ABERTA.value),
        pendentes=contar(Conversa.status == StatusConversa.PENDENTE.value),
        resolvidas_hoje=contar(
            Conversa.status == StatusConversa.RESOLVIDA.value, Conversa.resolvida_em >= inicio
        ),
        sem_atendente=contar(
            Conversa.atendente_id.is_(None), Conversa.status != StatusConversa.RESOLVIDA.value
        ),
        mensagens_hoje=mensagens_hoje,
        por_canal=por_canal,
        tempo_medio_primeira_resposta_seg=media,
    )
